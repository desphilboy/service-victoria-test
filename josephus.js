export class Josephus {

	constructor() {}

	static josephus(n, k, s = 1) {
		let r;
		if (k < 1) {
			return;
		}
		let circle = [];
		let sequence = [];
		let effectiveCount = n;
		for(let i=0; i < n; ++i) {
			circle.push(i+1);
		}
		let pos = s-1;

		while(circle.length > 1) {
			pos = (pos + k)
			if (pos >= circle.length) {
				pos = (pos % circle.length);
			} 
			r = circle.splice(pos - 1, 1)[0];
			pos = pos - 1;
			sequence.push(r);
		}
		sequence.push(circle[0]);
		return sequence;
	}
}
