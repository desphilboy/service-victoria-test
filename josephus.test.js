import { Josephus } from "./josephus.js";

describe("Josephus", () => {
  it("gives correct sequence for n=5, k=2", () => {
    expect(Josephus.josephus(5, 2)).toMatchSnapshot();
  });

  it("gives correct sequence for n=25, k=4", () => {
    expect(Josephus.josephus(25, 4)).toMatchSnapshot();
  });

  it("gives normal chain or k=1", () => {
    expect(Josephus.josephus(4, 1)).toMatchInlineSnapshot(`
      Array [
        1,
        2,
        3,
        4,
      ]
    `);
  });
  it("returns empty if n=0", () => {
    expect(Josephus.josephus(0, 4)).toMatchInlineSnapshot(`
      Array [
        undefined,
      ]
    `);
  });
});
