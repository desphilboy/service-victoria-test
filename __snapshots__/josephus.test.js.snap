// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Josephus gives correct sequence for n=5, k=2 1`] = `
Array [
  2,
  4,
  1,
  5,
  3,
]
`;

exports[`Josephus gives correct sequence for n=25, k=4 1`] = `
Array [
  4,
  8,
  12,
  16,
  20,
  24,
  3,
  9,
  14,
  19,
  25,
  5,
  11,
  18,
  1,
  10,
  21,
  6,
  17,
  7,
  23,
  15,
  22,
  2,
  13,
]
`;
